### PR TITLE
[Enhancement] Change catalog lock to RWLock to reduce lock competition

### LINKED ISSUE #4889: Optimize lock performance
## Enhancement
1.catalog change to readwritelock
2.reduce lock to table granularity

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4889

## Problem Summary(Required) ：

For catalog locks, there are currently three types of operations that require this lock to maintain.

1. Modification of DB meta data.
2. Operations on Table meta data.
3. Operations on FrontEnd meta data.

Among them, operations on DB rarely occur, and operations on Table are easy to occur, so this is a scenario with more reads and fewer writes. We should use readwritelock to maintain these operations. 
For FrontEnd, due to the relatively low cost, we use write lock maintenance, 
we use read locks for table create operations to improve lock concurrency, 
and this is also the first step in which lock granularity can be reduced to Table.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 22 / 75 (29.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|20|72|27.78%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|2|3|66.67%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 1]
```diff
@@ -295,11 +295,11 @@
...
     // Lock to perform atomic modification on map like 'idToDb' and 'fullNameToDb'.
     // These maps are all thread safe, we only use lock to perform atomic operations.
-    // Operations like Get or Put do not need lock.
+    // Operations like Get or Put do not need to lock.
     // We use fair ReentrantLock to avoid starvation. Do not use this lock in critical code pass
     // because fair lock has poor performance.
-    // Using QueryableReentrantLock to print owner thread in debug mode.
-    private QueryableReentrantLock lock;
+    // Using QueryableReentrantReadWriteLock to print owner thread in debug mode.
+    private QueryableReentrantReadWriteLock rwLock;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:302]: It's better to give the reason and performance test for this lock change.

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 2]
```diff
@@ -2698,21 +2725,21 @@ public void addCluster(Cluster cluster) {
     }
 
     public void replayCreateDb(Database db) {
-        tryLock(true);
+        tryReadLock(true);
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:2728]: CreateDb should use write lock

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 3]
```diff
@@ -773,19 +772,47 @@ private boolean tryLock(boolean mustLock) {
...
+            }
+        }
+    }
+
+    private void readUnlock() {
+        this.rwLock.readLock().unlock();
+    }
+
+    private void writeUnlock() {
+        if (this.rwLock.writeLock().isHeldByCurrentThread()) {
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:814]: Is this judge wisdom? Lock and unlock must appear in pairs, this should be guaranteed by the caller. I think this judge will hide some problems.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 22 / 75 (29.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|20|72|27.78%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|2|3|66.67%|

**Comment by Astralidea** [comment]
This close due to RWlock performance testing is not ideal.

**Pull request closed without merging**